pkg/lease/lease_test.go

```diff
@@ -0,0 +1,305 @@
+		option.WithoutAuthentication(),
+		option.WithGRPCDialOption(grpc.WithInsecure()),
+		option.WithGRPCDialOption(grpc.WithContextDialer(bufDialer)),
+	)
+	if err != nil {
+		t.Fatalf("creating metric client failed: %s", err)
+	}
+
+	var (
+		resource = &monitoredres_pb.MonitoredResource{
+			Type:   "prometheus_target",
+			Labels: map[string]string{"instance": "replica-key"},
+		}
+		intervals = newReplicaIntervals(leaseOpts, time.Now())
+		// Control number of concurrent replicas via channel.
+		replicaCh = make(chan struct{}, replicaCount)
+	)
+	hook := func(name string, start, end time.Time, valid bool) {
+		if valid {
+			intervals.add(name, start, end)
+		}
+	}
+	logf, err := ioutil.TempFile("", "lease")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer logf.Close()
+
+	t.Logf("log output at %s", logf.Name())
+
+	logger := log.NewLogfmtLogger(log.NewSyncWriter(logf))
+Loop:
+	for i := 0; ; i++ {
+		select {
+		case <-ctx.Done():
+			break Loop
+		case replicaCh <- struct{}{}:
+		}
+		// Keep producing new replica IDs. In practice IDs would be recycled but
+		// this makes it easier to debug history.
+		name := fmt.Sprintf("replica-%d", i)
+		rlogger := log.With(logger, "ts", log.DefaultTimestampUTC, "replica", name)
+
+		lease, err := New(rlogger, "test-project", client, resource, leaseOpts)
+		if err != nil {
+			t.Fatalf("Failed to create lease: %s", err)
+		}
+		lease.Register(func(start, end time.Time, ok bool) {
+			hook(name, start, end, ok)
+		})
+		// Make replicas die after a random amount of time unless a fixed lifetime is given.
+		if replicaLifetime == 0 {
+			replicaLifetime = time.Duration(rand.Intn(int(50 * leaseOpts.ExtensionPeriod)))
+		}
+
+		replicaCtx, cancel := context.WithTimeout(ctx, replicaLifetime)
+		go func() {
+			lease.Run(replicaCtx)
+			cancel()
+			// Free replica spot on exit.
+			<-replicaCh
+		}()
+	}
+
+	t.Log(intervals.intervals())	
+
+	if err := intervals.validate(); err != nil {
+		t.Fatal(err)
+	}
+	return intervals.intervals()
+}
+
+func TestLease(t *testing.T) {
+	opts := &Options{
+		UpdatePeriod:    10 * time.Millisecond,
+		RetryPeriod:     1 * time.Millisecond,
+		ExtensionPeriod: 30 * time.Millisecond,
+	}
+
+	t.Run("one-lease-interval", func(t *testing.T) {
+		// This test ensures that the lease sticks with a single replica the entire time if it doesn't die
+		// (and there are no transient errors preventing successful lease extension).
+		t.Parallel()
+		intervals := testLease(t, 5*time.Second, opts, 5, time.Hour, 0)
+
+		if len(intervals) != 1 {
+			t.Fatalf("expected exactly one interval with long-running replicas but got %d", len(intervals))
+		}
+	})
+	t.Run("two-intermittent-replicas", func(t *testing.T) {
+		// This test attempts to trigger a bad condition by having a lease change often across
+		// a high number of replicas.
+		t.Parallel()
+		testLease(t, 10*time.Second, opts, 2, 0, 0)
+	})
+	t.Run("many-intermittent-replicas", func(t *testing.T) {
```

Comment:
These last two tests are testing the same thing, with the latter being a stronger test. The `opts` between the two are the same, so the `UpdatePeriod` should be limiting the lease change frequency in both cases.